note/note_rfc6962.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package note
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	ct "github.com/google/certificate-transparency-go"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// RFC6962VerifierString creates a note style verifier string for use with NewRFC6962Verifier below.
+// logURL is the root URL of the log.
+// pubK is the public key of the log.
+func RFC6962VerifierString(logURL string, pubK crypto.PublicKey) (string, error) {
+	if !isValidName(logURL) {
+		return "", errors.New("invalid name")
+	}
+	pubSer, err := x509.MarshalPKIXPublicKey(pubK)
+	if err != nil {
+		return "", err
+	}
+	logID := sha256.Sum256(pubSer)
+	name := rfc6962LogName(logURL)
+	hash := rfc6962Keyhash(name, logID)
+	return fmt.Sprintf("%s+%08x+%s", name, hash, base64.StdEncoding.EncodeToString(append([]byte{algRFC6962STH}, pubSer...))), nil
+}
+
+// NewRFC6962Verifier creates a note verifier for Sunlight/RFC6962 checkpoint signatures.
+func NewRFC6962Verifier(vkey string) (note.Verifier, error) {
```

Comment:
Aha, I was trying to do this "clean room" from the spec :)

In general I really like the string vkey idea; it makes it very easy to poke through configs in existing code (e.g. the witness), so I'm all up for specifying for all the note-compatible verifiers we're cooking up!

On the extracting the timestamp, the way I was going to go was the same as I've done here for `TimestampCoSigv1`: https://github.com/transparency-dev/formats/blob/main/note/note_cosigv1.go#L116-L127

Passing the call-back in at verifier c'tor time means that in apps like the witness you can't really create verifiers out at "config parsing" time (with corresponding error handling there) and then pass them down into other funcs/modules without adding some other complexity around how you pass that callback in too. It also means that the verifier is no longer safe to use from multiple go-routines.

It "feels good" to have the timestamp extracted _during_ verification available, but OTOH, since `note.Note` already has a mechanism to convey which sigs (if any) were verified, I figured it's probably ok to have the application logic which relies on those signatures then go and explicitly extract the "extra" info present in them. Any malicious code which could modify/insert/remove entries from the `note.Sigs[]` immediately after the `note.Open()` returns can also probably just nobble the signature verification/timestamp extraction itself, or the callback and its side effects, etc.

wdyt?